### PR TITLE
SI-6956 determine switchability by type, not tree

### PR DESCRIPTION
**Pull request by adriaanm**
Constant folding will set the type of a constant tree
to `ConstantType(Constant(folded))`, while the tree
itself can be many different things (in casu, an Ident).

We used to look at the tree directly when deciding whether
to emit a switch. Now we look at the tree's type. Voilà.

review by @paulp

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -3494,8 +3494,10 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
       val alternativesSupported = true
       val canJump = true
 
-      object SwitchablePattern { def unapply(pat: Tree): Option[Tree] = pat match {
-        case Literal(const@Constant((_: Byte ) | (_: Short) | (_: Int  ) | (_: Char ))) =>
+      // Constant folding sets the type of a constant tree to `ConstantType(Constant(folded))`
+      // The tree itself can be a literal, an ident, a selection, ...
+      object SwitchablePattern { def unapply(pat: Tree): Option[Tree] = pat.tpe match {
+        case ConstantType(const@Constant((_: Byte ) | (_: Short) | (_: Int  ) | (_: Char ))) =>
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala]: - [x] `case ConstantType(const) if const.isIntRange`

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala]: Yeah! I didn't pour half my life into stupid old pattern matcher so you kids today could forget about isIntRange!

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 2]
```diff
@@ -3494,8 +3494,10 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
...
           Some(Literal(Constant(const.intValue))) // TODO: Java 7 allows strings in switches
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:3501]: Question: what is the relevance of java7 allowing strings in switches? It's a java-language level thing.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:3501]: I was assuming there would be support in bytecode as well. I guess I assumed wrong.

**Comment by adriaanm** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
The national institute of health applauds your code reuse.

**Pull request merged**